File: mvc1/_etc/Java/2422 한윤정이 이탈리아에 가서 아이스크림을 사먹는데.java
import java.util.*;

public class Main {
    
    static ArrayList<ArrayList<Integer>> graph = new ArrayList<ArrayList<Integer>>();

    static void Swap(int[] arr) {
         int temp = 0;
         temp = arr[0];
         arr[0] = arr[1];
         arr[1] = temp;
    }

    static boolean Next_Permutation(int[] arr, int end) {
        int i = end-1;
        int[] temp = new int[2];
        while(i > 0 && arr[i] <= arr[i-1]) --i;
            
        if(i == 0) return false;
            
        int j = end-1;
        while(arr[i-1] >= arr[j]) j -= 1;
        temp[0] = arr[i-1]; temp[1] = arr[j];
        Swap(temp);
        arr[i-1] = temp[0]; arr[j] = temp[1];
        j = end-1;
        while(i < j) {
            temp[0] = arr[i]; temp[1] = arr[j];
            Swap(temp);
            arr[i] = temp[0]; arr[j] = temp[1];
            i += 1; j-= 1;
        }
        return true;
    }
    
    static boolean dfs(int v, int w) {
        
        for(int to : graph.get(v)) {
            if(to == w) return false;
        }
        return true;
    }
    
    public static void main(String args[]) {
        for(int i = 0 ; i < 201; i++) graph.add(new ArrayList<Integer>());
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt(); int m = sc.nextInt();
        ArrayList<Integer> ice = new ArrayList<Integer> ();
        int[] index = new int[n];
        for(int i = 0; i < n; i++) ice.add(i+1);
        for(int i = 0 ; i < n-3; i++) index[i] = 0;
        for(int i = n-3 ; i < n; i++) index[i] = 1;
        
        for(int i = 0; i < m; i++) {
            int from = sc.nextInt(); int to = sc.nextInt();
            graph.get(from).add(to); graph.get(to).add(from);
        }
        
        int answer = 0;
        do {
            ArrayList<Integer> combi = new ArrayList<Integer> ();
            for(int i = 0 ; i < n; i++) {
                if (index[i] == 1) combi.add(ice.get(i));     
            }
            boolean can = true;
            for(int i = 0 ; i < 3; i++) {
                for(int j = i+1 ; j < 3; j++) {
                    
                    if(!dfs(combi.get(i), combi.get(j))) {
                        can = false; break;
                    }
                    
                }
                if(!can) break;
            }
            
            if(can) {
                answer++;
            }
        } while(Next_Permutation(index, n));
        
        System.out.println(answer);
    }
}

File: Beakjoon/_etc/Kotlin/2422 한윤정이 이탈리아에 가서 아이스크림을 사먹는데.kt
import java.util.*;


var graph = ArrayList<ArrayList<Int>>();

fun Swap(arr : IntArray) {
    var tmp: Int;
    tmp = arr[0];
    arr[0] = arr[1];
    arr[1] = tmp;
}

fun Next_Permutation(arr : IntArray, end : Int) : Boolean {
    var i = end-1;
    var temp = IntArray(2, {0});
    while(i > 0 && arr[i] <= arr[i-1]) --i;
        
    if(i == 0) return false;
        
    var j = end-1;
    while(arr[i-1] >= arr[j]) j -= 1;
    temp[0] = arr[i-1]; temp[1] = arr[j];
    Swap(temp);
    arr[i-1] = temp[0]; arr[j] = temp[1];
    j = end-1;
    while(i < j) {
        temp[0] = arr[i]; temp[1] = arr[j];
        Swap(temp);
        arr[i] = temp[0]; arr[j] = temp[1];
        i += 1; j-= 1;
    }
    return true;
}

fun dfs(v : Int, w : Int) : Boolean {
    for(to in graph.get(v)) {
        if(to == w) return false;
    }
    return true;
}

fun main(args: Array<String>) {
    val sc = Scanner(System.`in`);
    
    val n = sc.nextInt(); val m = sc.nextInt();
    for(i in 0..n) graph.add(ArrayList<Int>())
    for(i in 0 until m) {
        val f = sc.nextInt(); val t = sc.nextInt();
        graph.get(f).add(t); graph.get(t).add(f);
    }
    
    var index = IntArray(n, {0});
    for(i in n-3 until n) index[i] = 1;
    for(i in 0 until n-3) index[i] = 0;
    var ice = ArrayList<Int>();
    for(i in 0 until n) ice.add(i+1);
    var answer = 0;
    do {
        var combi = ArrayList<Int>();
        for(i in 0 until n) {
            if(index[i] == 1) combi.add(ice[i]);
        }
        var can = true;
        loop@ for(i in 0 until 3) {
            for(j in i+1 until 3) {
                if(!dfs(combi.get(i), combi.get(j))) {
                    can = false; break@loop;
                }
            }
        }
        
        if(can) answer++;
        
    
    } while(Next_Permutation(index, n));
    
    print(answer);
}
